html2text/html2text.1
.\" manpage for html2text
.\"
.TH html2text 1 2020\-04\-15
.SH NAME
html2text \- an advanced HTML\-to\-text converter
.SH SYNOPSIS
.B html2text -help
.br
.B html2text -version
.br
.B html2text
[
.B \-unparse
|
.B \-check
] [
.B \-debug\-scanner
] [
.B \-debug\-parser
] [
.B \-rcfile
.I path
] [
.B \-width
.I width
] [
.B \-o
.I output-file
] [
.B \-nobs
] [
.B \-from_encoding
.I encoding
] [
.B \-to_encoding
.I encoding
] [
.B \-ascii
] [
.IR input-file " ..."
]
.SH DESCRIPTION
.B html2text
reads HTML documents from the
.IR input-file s,
formats each of them into a stream of UTF-8 encoded characters,
and writes the result to standard output (or into
.IR output-file ,
if the
.B -o
command line option is used).
.P
If no
.IR input-file s
are specified on the command line,
.B html2text
reads from standard input. A dash as the
.I input-file
is an alternate way to specify standard input.
.P
.B html2text
understands all HTML 3.2 constructs, but can render only part of them due to
the limitations of the text output format. However, the program attempts to
provide good substitutes for the elements it cannot render.
.B html2text
parses HTML 4 input, too, but not always as successful as other HTML
processors. It also accepts syntactically incorrect input, and attempts to
interpret it "reasonably".
.P
The way
.B html2text
formats the HTML documents is controlled by formatting properties read
from an RC file.
.B html2text
attempts to read
.I $HOME/.html2textrc
(or the file specified by the
.B -rcfile
command line option); if that file cannot be read,
.B html2text
attempts to read
.IR /etc/html2textrc .
If no RC file can be read (or if the RC file does not override all
formatting properties), then "reasonable" defaults are assumed. The
RC file format is described in the
.BR html2textrc (5)
manual page.
.SH OPTIONS
.TP
.B \-ascii
By default,
.B html2text
uses
.B UTF-8
for its output. Specifying this option, plain
.B ASCII
is used instead. Non\-ASCII characters are rendered via iconv's
transliteration feature.  As such this option is an alias for
.B -to_encoding
.IR ASCII//TRANSLIT .
.TP
.BI "\-from_encoding " encoding
Sets the encoding of the input file or stream to the given encoding.  By
default,
.B html2text
uses ISO-8859-1 encoding for reading.  You probably always want to
override this default using this option.  See 
.B iconv -l
for a list of supported encodings.
.TP
.BI "\-to_encoding " encoding
Use the given encoding while writing output, instead of UTF-8.  The same
set of encodings as for the
.B \-from_encoding
option can be used.
.TP
.B \-check
This option is for diagnostic purposes: The HTML document is only parsed and
not processed otherwise. In this mode of operation,
.B html2text
will report on parse errors and scan errors, which it does not in other modes
of operation. Note that parse and scan errors are not fatal for
.BR html2text ,
but may cause mis-interpretation of the HTML code and/or portions of the
document being swallowed.
.TP
.B \-debug\-parser
Let
.B html2text
report on the tokens being shifted, rules being applied, etc., while scanning
the HTML document. This option is for diagnostic purposes.
.TP
.B \-debug\-scanner
Let
.B html2text
report on each lexical token scanned, while scanning the HTML document. This
option is for diagnostic purposes.
.TP
.B \-help
Print command line summary and exit.
.TP
.B \-nobs
By default,
.B html2text
renders underlined letters with sequences like "underscore-backspace-character"
and boldface letters like "character-backspace-character", which works fine
when the output is piped into
.BR more (1),
.BR less (1),
or similar. For other applications, or when redirecting the output into a file,
it may be desirable not to render character attributes with such backspace
sequences, which can be accomplished with this command line option.
.TP
.BI \-o " output\-file"
Write the output to
.I output\-file
instead of standard output. A dash as the
.I output\-file
is an alternate way to specify the standard output.
.TP
.BI \-rcfile " path"
Attempt to read the file specified in
.I path
as RC file.
.TP
.B \-links
Tags all external links with a number between brackets
.RI "([" num "]),"
and produces a numbered list at the end of the document with all link
targets.
.TP
.B \-unparse
This option is for diagnostic purposes: Instead of formatting the parsed document,
generate HTML code, that is guaranteed to be syntactically correct. If
.B html2text
has problems parsing a syntactically incorrect HTML document, this option
may help you to understand what
.B html2text
thinks that the original HTML code means.
.TP
.B \-version
Print program version and exit.
.TP
.BI \-width " width"
By default,
.B html2text
formats the HTML documents for a screen width of 79 characters. If redirecting
the output into a file, or if your terminal has a width other than 80
characters, or if you just want to get an idea how
.B html2text
deals with large tables and different terminal widths, you may want to specify
a different
.IR width .
.SH FILES
.TP
.I /etc/html2textrc
System wide parser configuration file.
.TP
.I $HOME/.html2textrc
Personal parser configuration file, overrides the system wide values.
.SH "CONFORMING TO"
.B "HTML 3.2"
(HTML 3.2 Reference Specification \- http://www.w3.org/TR/REC-html32).
.SH RESTRICTIONS
.B html2text
was written to convert HTML 3.2 documents. When using it with HTML 4 or
even XHTML 1 documents, some constructs present only in these HTML versions
might not be rendered.
.SH AUTHOR
.B html2text
was written up to version 1.2.2 by Arno Unkrig <arno@unkrig.de>
for GMRS Software GmbH, Unterschleissheim.
.P
Up to version 1.3.2a, the maintainer was
Martin Bayer <mbayer@zedat.fu-berlin.de>.
.br
.P
This version of
.B html2text
comes from
https://github.com/grobian/html2text.  Please use the GitHub page to
file issues and improvements.
.SH SEE ALSO
.BR html2textrc (5),
.BR less (1),
.BR more (1)


libebu/NEWS
バージョン 4.4.3 における主な変更点:

* バージョン4.4.2で、共有ライブラリのバージョンが正しくなかったため、
  本バージョンで修正した。
* 細かな不具合を修正した。

バージョン 4.4.2 における主な変更点:

* 図版メニューに関する処理で、初期化処理が一部抜けていたのを修正した。
* ebzip 伸長時に、正常終了時も終了コードが1になってしまう不具合を修正
  した。
* ebzip テストモード時も、ファイルのコピーが実行されてしまう不具合を
  修正した。
* ebzip のメモリ開放忘れを直した。

バージョン 4.4.1 における主な変更点:

* 画像、音声データを扱う際に、データが格納されているファイルの大きさが
  2GB を正しく扱えなかったので修正した。

バージョン 4.4 における主な変更点:

* 2GB を超える大きさのファイルを扱えるようにした。ただし、システム側で
  対応していることが前提条件。
* EBZIP形式の圧縮ファイルで、4294967296〜1099511627775バイト (4GB〜1TB)
  のファイルの表現方法を規定した。ebzip, ebunzip はこの規定に沿って圧縮
  および伸張を行なう。
* ebzip, ebunzip がディレクトリ作成に失敗したとき、何もエラーメッセージ
  を出さずに終了する不具合を修正した。

バージョン 4.3.4 における主な変更点:

* 環境によっては、ebinfo の出力する副本の題名が文字化けするのを修正した。
* 注: ライブラリ本体には変更がないので、共有ライブラリのバージョンは更新
  してしない。

バージョン 4.3.3 における主な変更点:

* Cygwin, MinGW で DLL を作成できるようにした。
* Windows で ebzipinfo が異常終了する不具合を修正した。

バージョン 4.3.2 における主な変更点:

* バージョン 4.3.1 で、共有ライブラリの番号付けが間違っていたので修正した。
* ebinfo の日本語メッセージに、空白文字が抜けているものがあったので修正した。

バージョン 4.3.1 における主な変更点:

* 動画ファイル名に英大文字が使われると、動画データを読み込めない不具合
  を修正した。
* 電子ブックリーダー Panasonic KX-EBP2 付属品に対応した。
* ebnet でバッファオーバーランが起きる不具合を修正した。

バージョン 4.3 における主な変更点:

* EPWING で複数の副本がディレクトリを共有している場合、正しくアクセス
  できない不具合を修正した。
* `ebunzip --test' が、電子ブックの本文ファイルを上書きしてしまう不具合
  を修正した。
* ebappendix が、`catalogs.app' を読み込んでも電子ブック形式の appendix
  を生成してしまう不具合を修正した。
* 参照先付きカラー図版に対応した。
  API として、eb_have_image_menu() と eb_image_menu() が追加された。
  テキストのフックとして、EB_HOOK_BEGIN_IMAGE_PAGE, EB_HOOK_END_IMAGE_PAGE,
  EB_HOOK_BEGIN_CLICKABLE_AREA, EB_HOOK_END_CLICKABLE_AREA の 4つが追加
  された。

バージョン 4.2.2 における主な変更点:

* 4.2.1 で混入した、一部の EPWING が正しく扱えない不具合を修正した。
* eb.conf の EBCONF_ENABLE_NLS の値が正しくセットされない不具合を修正
  した。
* 音声ファイルの取り出しがうまくいかない不具合を終始した。
* 一部の Linux 環境で configure 時に NLS (Native/National Language
  Support) が有効にならない不具合を修正した。
* 細かな不具合をいくつか修正した。

バージョン 4.2.1 における主な変更点:

* 不具合をいくつか修正した。

バージョン 4.2 における主な変更点:

* クロス検索に対応した。
  関数 eb_have_cross_search(), eb_search_cross() をライブラリに追加した。
  また、ebinfo は、クロス検索の情報を出力するようになった。
* 動画ファイル名の絶対パス名を返す関数 eb_compose_movie_path_name() を
  ライブラリに追加した。
* ebrefile は、カタログファイルを上書きするようにした。
  上書き前のカタログファイルは、".old" という接尾子の付いたファイル名で
  保存される。
* zlib を添付しないようにした。
  configure から --with-included-zlib オプションを削除した。
* 不具合をいくつか修正した。

バージョン 4.1.3 における主な変更点:

* バージョン 4.1.2 で、共有ライブラリの番号付けが間違っていたので修正した。
* `make install' が失敗する不具合を修正した。

バージョン 4.1.2 における主な変更点:

* 特定の書籍で、EB ライブラリが書籍に載っている単語を検索できない
  不具合を修正した。
* FreeBSD-5.3-RELEASE で make が失敗する不具合を修正した。
* Windows 環境において、ebzip コマンドの起動時のコマンド名 (ebunzip.exe,
  ebzipinfo.exe など) をもとに、正しく動作を決定できない不具合を修正
  した。

バージョン 4.1.1 における主な変更点:

* ユーティリティコマンドで、引数を取る長いオプション (例: ebzip の
  --level) に対して引数を渡さないと、コマンドが異常終了する不具合を
  修正した。
* eb_write_text_byte2() がおかしなデータを書き込む不具合を修正した。
* ドキュメントの HTML ファイルのスタイルを変更した。

バージョン 4.1 における主な変更点:

* BSD ライセンスを採用するようにした。
* ebfont が PNG 形式の画像ファイルを生成できるようにした。
  ebfont の --image-format オプションに `png' を指定できるようにした。
* ライブラリに eb_narrow_font_png_size(), eb_wide_font_png_size(),
  eb_bitmap_to_png() を加えた。
* eb_bitmap_to_xbm(), eb_bitmap_to_xpm(), eb_bitmap_to_gif(),
  eb_bitmap_to_bmp() は EB_Error_Code を返すようにた。
* ライブラリが、非圧縮ブロックデータのキャッシュをより効率的に行う
  ようにした。
* `ebzip' が START/HONMON ファイルの部分圧縮に対応した。
  これにより、前方一致検索が速くなった。
* 日本語ドキュメントを HTML 形式で配布するようにした。
* 英語のドキュメントを削除した。

バージョン 4.0.1 における主な変更点:

* ヘッダファイル内の EB_P, ZIO_P マクロの定義を修正した。
* `ebzipinfo' は `--skip-content movie' オプションを指定しない限り、
  動画ファイルの情報も表示するようにした。
* `ebzip', `ebunzip' の進行状況メッセージは、標準出力ではなく標準エラー
  出力に表示するようにした。

バージョン 4.0 における主な変更点:

* EB_BookList 関連の API を改良した。
* EB_ERR_UNBOUND_BOOKLIST, EB_ERR_NO_SUCH_BOOK という 2 つのエラー
  コードを追加した。
* `eb3.m4' を `eb4.m4' に改名した。
* `configure' に `--enable-ebnet' オプションを追加した。
* `eb/flags.h' をインストールするようにした。
  このファイルは、`eb/defs.h' から自動的に include される。
* 重大および軽度の不具合をいくつか修正した。

バージョン 4.0beta3 における主な変更点:

* EUC-JP 以外の日本語のエンコーディングに対応した。
* Windows 対応のためのコードをいくつか付加した。
  Windows 上で EB ライブラリのビルドを行うには、`eb-win32' パッケージ
  を使用して下さい。
* `ebstopcode' に `--text-position' (`-p') オプションを追加した。

バージョン 4.0beta2 における主な変更点:

* 検索メソッドの重大な不具合を修正した。

バージョン 4.0beta1 における主な変更点:

* `ebstopchk' を `ebstopcode' に改名した。
  コマンドの `--stop-code' オプションも `--code' に改名した。
* `ebappendix' の `--case', `--suffix' オプションを削除した。
* `ebappendix' の日本語版ドキュメントを用意した。
* `ebinfo' に `--book-list' オプションを追加した。
* `ebdump', `ebfixlog' を `misc' サブディレクトリに入れた。
* 不具合をいくつか修正した。

バージョン 4.0beta0 における主な変更点:

* 遠隔アクセスに対応した。
* `ebstopchk' コマンドを追加した。

バージョン 3.3.2 における主な変更点:

* 共有ライブラリのバージョン番号を修正した。
* 「日英仏辞典」(YRRS-059) に対応した。

バージョン 3.3.1 における主な変更点:

* EPWING 版「ジーニアス英和大辞典」に対応した。
* 複合検索の処理の不具合を修正した。
* デバッグモードの不具合を修正した。

バージョン 3.3 における主な変更点:

* eb_read_binary_mpeg() の不具合を修正した。

バージョン 3.3beta3 における主な変更点:

* 複合検索を修正した。
* 軽度の不具合をいくつか修正した。

バージョン 3.3beta2 における主な変更点:

* 重大および軽度の不具合をいくつか修正した。
* `eb3.m4' は `$(datadir)/aclocal' にインストールするようにした。

バージョン 3.3beta1 における主な変更点:

* 不具合をいくつか修正した。

バージョン 3.3beta0 における主な変更点:

* EPWING のファイル名や圧縮タイプを見分けるコードを大幅に書き改めた。
* 関数 eb_multi_title() を追加した。
  この関数は、特定の複合検索の題名を取得する。この関数の使用によって、
  `ebinfo' コマンドで `--multi-search' (`-m') を指定した際に、複合検索
  の題名を表示するようになった。
* `ebrefile' の不具合を修正した。
* インライン画像に対応するため、テキストデータへのフックを 2 種類追加
  した。

バージョン 3.2.3 における主な変更点:

* PCM 音声データの処理における不具合を修正した。
* テキストのフックを 2 種類追加した。

バージョン 3.2.2 における主な変更点:

* appendix の処理における不具合を修正した。
* gettext のメッセージカタログを修正した。

バージョン 3.2.1 における主な変更点:

* テキストの読み込みルーチンを修正した。
* zlib-1.1.4 を取り込んだ。

バージョン 3.2 における主な変更点:

* 特になし。

バージョン 3.2beta3 における主な変更点:

* インストールの際に起こる不具合をいくつか修正した。

バージョン 3.2beta2 における主な変更点:

* 環境変数 EB_DEBUG が定義されていると、EB ライブラリがデバッグ
  メッセージを出力するようにした。
  `configure' の `--enable-debug' オプションを削除した。
* EBZIP 圧縮形式データの伸長における致命的な不具合を修正した。

バージョン 3.2beta1 における主な変更点:

* `ebzip' の最高圧縮レベルを 3 から 5 に変更した。
* 不具合をいくつか修正した。

バージョン 3.2beta0 における主な変更点:

* eb_backward_text() を実装し、eb_forward_text() を改良した。
* EB ライブラリのプログラミングドキュメント (日本語版) を復活させた。
* サンプルプログラムを復活させた。

バージョン 3.1 における主な変更点:

* 圧縮された S-EBXA へのサポートにおける不具合を修正した。

バージョン 3.1beta1 における主な変更点:

* デバッグコードを付加した。
  このコードは `configure --enable-debug' とすると、有効になる。
* eb_initialize_all_subbooks() を eb_load_all_subbooks() に改名した。
  eb_initialize_all_appendix_subbooks() も eb_load_all_appendix_subbooks()
  に改名した。

html2text/html2text.cpp

/***************************************************************************/

/*
 * Portions Copyright (c) 1999 GMRS Software GmbH
 * Carl-von-Linde-Str. 38, D-85716 Unterschleissheim, http://www.gmrs.de
 * All rights reserved.
 *
 * Author: Arno Unkrig <arno@unkrig.de>
 */

/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License in the file COPYING for more details.
 */

/***************************************************************************/

/*
 * Changes to version 1.2.2 were made by Martin Bayer <mbayer@zedat.fu-berlin.de>
 * Dates and reasons of modifications:
 * Fre Jun  8 19:00:26 CEST 2001: new image handling
 * Thu Oct  4 21:42:24 CEST 2001: ported to g++ 3.0, bugfix for '-' as synonym for STDIN
 * Mon Jul 22 13:48:26 CEST 2002: Made finaly reading from STDIN work.
 * Sat Sep 14 15:04:09 CEST 2002: Added plain ASCII output patch by Bela Lubkin
 * Wed Jul  2 22:08:45 CEST 2003: ported to g++ 3.3
 */

/***************************************************************************/

#include "HTMLControl.h"
#include "HTMLDriver.h"

#ifndef HTML2TEXT_LIBS
#include "format.h"
#include "iconvstream.h"
#include <stdlib.h>

#define stringify(x) stringify2(x)
#define stringify2(x) #x

/* ------------------------------------------------------------------------- */

static const char *usage = "\
Usage:\n\
  html2text -help\n\
  html2text -version\n\
  html2text [ -unparse | -check ] [ -debug-scanner ] [ -debug-parser ] \\\n\
     [ -rcfile <file> ] [ -width <w> ] [ -nobs ] [ -links ]\\\n\
     [ -from_encoding ] [ -to_encoding ] [ -ascii ]\\\n\
     [ -o <file> ] [ <input-file> ] ...\n\
Formats HTML document(s) read from <input-file> or STDIN and generates ASCII\n\
text.\n\
  -help          Print this text and exit\n\
  -version       Print program version and copyright notice\n\
  -unparse       Generate HTML instead of ASCII output\n\
  -check         Do syntax checking only\n\
  -debug-scanner Report parsed tokens on STDERR (debugging)\n\
  -debug-parser  Report parser activity on STDERR (debugging)\n\
  -rcfile <file> Read <file> instead of \"$HOME/.html2textrc\"\n\
  -width <w>     Optimize for screen widths other than 79\n\
  -nobs          Do not render boldface and underlining (using backspaces)\n\
  -links         Generate reference list with link targets\n\
  -from_encoding Treat input encoded as given encoding\n\
  -to_encoding   Output using given encoding\n\
  -ascii         Use plain ASCII for output instead of UTF-8\n\
                 alias for: -to_encoding ASCII//TRANSLIT \n\
  -o <file>      Redirect output into <file>\n\
";

int
main(int argc, char **argv)
{
    if (argc == 1 || (argc == 2 && !strcmp(argv[1], "-help"))) {
        std::cout
            << "This is html2text, version " stringify(VERSION) << std::endl
            << std::endl
            << usage;
        exit(0);
    }

    if (argc == 2 && !strcmp(argv[1], "-version")) {
        std::cout
            << "This is html2text, version " stringify(VERSION) << std::endl
            << std::endl
            << "The latest version can be found at "
            << "https://github.com/grobian/html2text" << std::endl
            << std::endl
            << "This program is distributed in the hope that it will "
            << "be useful, but WITHOUT" << std::endl
            << "ANY WARRANTY; without even the implied warranty of "
            << "MERCHANTABILITY or FITNESS" << std::endl
            << "FOR A PARTICULAR PURPOSE. See the GNU General Public "
            << "License for more details." << std::endl
            << std::endl;
        exit(0);
    }

    int mode = HTMLDriver::PRINT_AS_ASCII;
    bool debug_scanner = false;
    bool debug_parser = false;
    const char *home = getenv("HOME");
    const char *rcfile = NULL;
    int width = 79;
    const char *output_file_name = "-";
    bool use_backspaces = true;
    bool enable_links = false;
    const char *from_encoding = NULL;
    const char *to_encoding = NULL;
    const char *widthstr = NULL;
    const char **extarg = NULL;

    int i;
    for (i = 1; i < argc && argv[i][0] == '-' && argv[i][1]; ++i) {
        const char *arg = argv[i];

        if (!strcmp(arg, "-unparse")) {
            mode = HTMLDriver::UNPARSE;
        } else if (!strcmp(arg, "-check")) {
            mode = HTMLDriver::SYNTAX_CHECK;
        } else if (!strcmp(arg, "-debug-scanner")) {
            debug_scanner = true;
        } else if (!strcmp(arg, "-debug-parser")) {
            debug_parser = true;
        } else if (!strcmp(arg, "-rcfile")) {
            extarg = &rcfile;
        } else if (!strcmp(arg, "-links")) {
            enable_links = true;
        } else if (!strcmp(arg, "-width")) {
            extarg = &widthstr;
        } else if (!strcmp(arg, "-o")) {
            extarg = &output_file_name;
        } else if (!strcmp(arg, "-nobs")) {
            use_backspaces = false;
        } else if (!strcmp(arg, "-from_encoding")) {
            extarg = &from_encoding;
        } else if (!strcmp(arg, "-to_encoding")) {
            extarg = &to_encoding;
        } else if (!strcmp(arg, "-ascii")) {
            to_encoding = "ASCII//TRANSLIT";  /* create things like (c) */
        } else {
            std::cerr
                << "Unrecognized command line option \""
                << arg
                << "\", try \"-help\"."
                << std::endl;
            exit(1);
        }

        if (extarg != NULL) {
            if (++i >= argc) {
                std::cerr
                    << "Option \"" << arg << "\" needs an additional "
                    << "argument."
                    << std::endl;
                exit(1);
            }
            *extarg = argv[i];

            /* handle the only integer argument inline */
            if (extarg == &widthstr) {
                int nwidth = atoi(widthstr);
                if (nwidth > 10) {
                    width = nwidth;
                } else {
                    std::cerr
                        << "width '" << nwidth << "' invalid, must be >10"
                        << std::endl;
                    exit(1);
                }
            }
            extarg = NULL;
        }
    }
    if (i > argc) {
        std::cerr
            << "Error: Required parameter after \""
            << argv[argc - 1]
            << "\" missing."
            << std::endl;
        exit(1);
    }

    /* historical default */
    if (from_encoding == NULL)
        from_encoding = "ISO-8859-1";

    /* this is probably the output we want on 99% of all terminals */
    if (to_encoding == NULL)
        to_encoding = "UTF-8";

    const char *const *input_files;
    int number_of_input_files;

    if (i >= argc) {
        static const char *const x = "-";
        input_files = &x;
        number_of_input_files = 1;
    } else {
        input_files = argv + i;
        number_of_input_files = argc - i;
    }

    {
        std::ifstream ifs;

        if (rcfile == NULL && home != NULL)
            rcfile = (string(home) + "/.html2textrc").c_str();
        if (rcfile != NULL)
            ifs.open(rcfile);
        if (rcfile == NULL || !ifs.rdbuf()->is_open())
            ifs.open("/etc/html2textrc");
        if (ifs.rdbuf()->is_open())
            Formatting::loadProperties(ifs);
    }

    /*
     * Set up printing.
     */
    //Area::use_backspaces = use_backspaces;

    iconvstream is(use_backspaces);

    is.open_os(output_file_name, to_encoding);
    if (!is.os_open()) {
        std::cerr
            << "Could not open output file \""
            << output_file_name
            << "\": "
            << is.open_error_msg()
            << std::endl;
        exit(1);
    }

    for (i = 0; i < number_of_input_files; ++i) {
        const char *input_file = input_files[i];

        if (number_of_input_files != 1)
            is << "###### " << input_file << " ######" << endl_char;

        is.open_is(input_file, from_encoding);
        if (!is.is_open()) {
            std::cerr
                << "Opening input file \""
                << input_file
                << "\": "
                << is.open_error_msg()
                << std::endl;
            exit(1);
        }

        HTMLControl control(is, mode, debug_scanner, input_file);
        HTMLDriver driver(control, is, enable_links,
                width, mode, debug_parser);

        if (driver.parse() != 0)
            exit(1);
    }

    is.close();

    return 0;
}

#else
#include "html2text.h"

bool html_to_text(h2t_iostream &h2tio, int width, bool enable_links)
{
    const int mode = HTMLDriver::PRINT_AS_ASCII;

    HTMLControl control(h2tio, mode, false);
    HTMLDriver driver(control, h2tio, enable_links, width, mode, false);

    return (driver.parse() == 0);
}
#endif

h2t_iostream &h2t_iostream::operator<<(const char *inp)
{
    write(inp, strlen(inp));
    return *this;
}

h2t_iostream &h2t_iostream::operator<<(const string &inp)
{
    write(inp.c_str(), inp.length());
    return *this;
}

h2t_iostream &h2t_iostream::operator<<(char inp)
{
    write(&inp, inp == '\0' ? 0 : 1);
    return *this;
}

/* See istr::append_utf8char */
h2t_iostream &h2t_iostream::operator<<(unsigned int inp)
{
    unsigned char point = 0;
    char fc = 0;
    do {
        fc = ((unsigned)inp >> (24 - 8 * point)) & 0xff;
        ++point;
    }while(!fc && point < 4);

    write(&fc, fc == '\0' ? 0 : 1);

    if ((fc >> 7) & 1) {
        char single = 0;
        while(( single = (inp >> (24 - 8 * point)) & 0xff ) && point++ < 4)
            write(&single, 1);
    }
    return *this;
}

/* See istr::from_chars */
h2t_iostream &h2t_iostream::operator>>(unsigned int &op)
{
    op = getc();

    unsigned char fc = op & 0xFF;

    if ((fc >> 7) & 1) {
        unsigned char nextpoint = 1;

        /* we assume iconv produced valid UTF-8 here */
        while ( ((fc >> (7 - nextpoint)) & 1) && nextpoint++ < 4 )
        {
            op <<= 8;
            op |= (getc() & 0xFF);
        }
    }

    return *this;
}

/* ------------------------------------------------------------------------- */
